### PR TITLE
flock is hard, even (especially?) when wrapped in a fluffy API

### PR DESCRIPTION
**Pull request by jacobstanley**
Create the lock file before trying to lock on it.

The [filelock](https://hackage.haskell.org/package/filelock) package is designed in such a way that if the file doesn't exist, there will be a race and two separate fds can be created for the same filename. This means that both processes / threads successfully obtain the lock.

/cc @markhibberd @nhibberd @thumphries

### CODE REVIEW

File: src/Mafia/Install.hs [hunk 1]
```diff
@@ -446,13 +447,29 @@ lockPackage :: PackageEnv -> Package -> Flavour -> EitherT InstallError IO FileL
...
+      OpenFileFlags {
+          append = False
+        , exclusive = False
+        , noctty = False
+        , nonBlock = False
+        , trunc = False
+        }
+  in
+    bracket (openFd path ReadOnly (Just 0o666) flags) closeFd $ \_ ->
+      pure ()
```
Comment (erikd-ambiata) [inline src/Mafia/Install.hs:471]: Use `const` maybe?

**Comment by erikd-ambiata** [comment]
:+1: on :green_apple:

**Comment by olorin** [comment]
Explanatory comment would be good, otherwise :+1:.

**Comment by thumphries** [comment]
... that's quite bad, defeats the whole premise of FileLock. Anything we can do about it upstream?

**Comment by jacobstanley** [comment]
> ... that's quite bad, defeats the whole premise of FileLock.

That's what I thought!

Yeah it should be fixable upstream, I think we just need to ensure the file exists before we get the fd to it.

Unix experts, does it make sense that two parallel calls to `open` can return different a fd?

**Comment by olorin** [comment]
> Unix experts, does it make sense that two parallel calls to open can return different a fd?

Two parallel calls to open will return different fds. Not sure how it could be otherwise?

**Comment by jacobstanley** [comment]
> Two parallel calls to open will return different fds. Not sure how it could be otherwise?

Sorry I don't mean fds, I mean inodes

**Comment by olorin** [comment]
> Sorry I don't mean fds, I mean inodes

Makes sense to me - `open(3)[thread1]` then `unlink(3)`, `rename(3)` and then `open(3)[thread2]`, unless I'm misunderstanding.

**Comment by jacobstanley** [comment]
No renaming or unlinking, I just mean both threads call `open(3)` at the same time with the same path and the file does not exist yet

**Comment by markhibberd** [comment]
Mmm, yeh, I think my comment was wrong, there is something else subtle going on here that I am missing.

**Comment by jacobstanley** [comment]
I could always switch to Max's [functor lock](https://github.com/ambiata/chum/blob/master/chum-ingest/main/Main.hs#L245-L271)

**Comment by olorin** [comment]
> No renaming or unlinking, I just mean both threads call open(3) at the same time with the same path and the file does not exist yet

This can create two inodes if it's not opened with `O_EXCL` as well as `O_CREAT` - atomicity is guaranteed with `O_EXCL` but not otherwise afaik.

**Comment by markhibberd** [comment]
Yeh, so that is one hole, the other is this pleasantly worded piece (http://linux.die.net/man/2/flock):

> An attempt to lock the file using one of these file descriptors may be denied by a lock that the calling process has already placed via another descriptor.

Hanging a lot on _may_. 

So horrid, complicated, but only real path might be touch + internal process lock as well.

**Comment by jacobstanley** [comment]
Closing in favour of https://github.com/ambiata/mafia/pull/119

**Pull request closed without merging**